### PR TITLE
#MySql Task. Arseni Lamihov

### PR DESCRIPTION
**Pull request by Arseni-Lamihov**
https://travis-ci.com/github/Arseni-Lamihov/db-assignments/builds/178464257

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -67,7 +81,16 @@ async function task_1_3(db) {
...
 async function task_1_4(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+    SELECT 
+        CustomerID as "Customer Id",
+        COUNT(*) as "Total number of Orders",
+        ROUND(COUNT(*)/(SELECT COUNT(*) FROM Orders)*100, 5) as "% of all orders"
+    FROM Orders 
+    GROUP BY CustomerID 
+    ORDER BY COUNT(*)/(SELECT COUNT(*) FROM Orders)*100 DESC, CustomerID;
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not duplicate formula

File: task/sql-tasks.js [hunk 2]
```diff
@@ -91,7 +123,17 @@ async function task_1_5(db) {
...
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+    SELECT 
+        P.ProductName,
+        C.CategoryName,
+        S.CompanyName as "SupplierCompanyName"
+    FROM Products as P
+    JOIN Suppliers as S ON S.SupplierID=P.SupplierID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: join is alias for inner join, but it's better to write inner join explicitly

File: task/sql-tasks.js [hunk 3]
```diff
@@ -177,7 +268,15 @@ async function task_1_12(db) {
...
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            COUNT(ProductID) as "TotalOfCurrentProducts",
+            COUNT(CASE
+                    WHEN Discontinued = 1 
+                      THEN 1 END ) as "TotalOfDiscontinuedProducts"           
+        FROM Products;
```
Comment (AlexMokin) [inline task/sql-tasks.js:277]: from-clause is optional.

 it's better to write this query like this

```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 4]
```diff
@@ -199,7 +307,24 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            SUM(MONTH(OrderDate) = 1) as January,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count to show intent explicitly

File: task/sql-tasks.js [hunk 5]
```diff
@@ -223,7 +356,16 @@ async function task_1_16(db) {
  *
  */
 async function task_1_17(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+          SELECT
+              C.CategoryName,
+              AVG(P.UnitPrice) AS "AvgPrice"
+          FROM Categories as C
+          JOIN Products as P ON C.CategoryID = P.CategoryID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 6]
```diff
@@ -247,7 +397,19 @@ async function task_1_18(db) {
...
  */
 async function task_1_19(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            C.CustomerID,
+            C.CompanyName,
+            SUM(UnitPrice * Quantity) AS 'TotalOrdersAmount, $'
+        FROM Customers as C
+        JOIN Orders as O ON O.CustomerID = C.CustomerID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 7]
```diff
@@ -259,7 +421,19 @@ async function task_1_19(db) {
...
  */
 async function task_1_20(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            E.EmployeeID as "EmployeeID",
+            CONCAT(E.FirstName, " ", E.LastName) as "Employee Full Name",
+            SUM(OD.UnitPrice * OD.Quantity) as "Amount, $"
+        FROM Employees as E
+        JOIN Orders as O ON O.EmployeeID = E.EmployeeID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 8]
```diff
@@ -280,30 +463,62 @@ async function task_1_21(db) {
...
+  //                             JOIN Customers as CNew ON ONew.CustomerID = CNew.CustomerID
+  // throw new Error("Not implemented");
 }
 
 module.exports = {
-    task_1_1: task_1_1,
-    task_1_2: task_1_2,
-    task_1_3: task_1_3,
-    task_1_4: task_1_4,
-    task_1_5: task_1_5,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not modify original code

File: task/sql-tasks.js [hunk 9]
```diff
@@ -280,30 +463,62 @@ async function task_1_21(db) {
...
+
+
+            
+  `);
+  return result[0];
+
+  
+  // FROM Customers c1
+  // JOIN Orders o2 ON o2.CustomerID = c1.CustomerID 
+  // JOIN OrderDetails od2 ON o2.OrderID = od2.OrderID )
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not add unnecessary code

File: task/sql-tasks.js [hunk 10]
```diff
@@ -280,30 +463,62 @@ async function task_1_21(db) {
...
+        FROM Customers as C
+        JOIN Orders as O ON O.CustomerID = C.CustomerID        
+        JOIN OrderDetails as OD ON OD.OrderID = O.OrderID
+        JOIN Products as P ON OD.ProductID = P.ProductID
+        WHERE OD.UnitPrice = (SELECT
+                                MAX(ODNew.UnitPrice)
+                              FROM Customers as CNew
+                              JOIN Orders as ONew ON ONew.CustomerID = CNew.CustomerID 
+                              JOIN OrderDetails as ODNew ON ONew.OrderID = ODNew.OrderID 
+                              WHERE CNew.CompanyName = C.CompanyName)
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by Arseni-Lamihov** [comment]
https://travis-ci.com/github/Arseni-Lamihov/db-assignments/jobs/379069901

File: task/sql-tasks.js [hunk 11]
```diff
@@ -91,7 +123,17 @@ async function task_1_5(db) {
...
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+    SELECT 
+        ProductName,
+        Categories.CategoryName,
+        Suppliers.CompanyName as "SupplierCompanyName"
+    FROM Products 
+    JOIN Suppliers  ON Suppliers.SupplierID = Products.SupplierID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use inner join explicitly.

File: task/sql-tasks.js [hunk 12]
```diff
@@ -177,7 +268,13 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            (COUNT(ProductID)) as 'TotalOfCurrentProducts',
+            (COUNT(CASE WHEN Discontinued = 1 THEN 1 END )) as 'TotalOfDiscontinuedProducts'         
+        FROM Products;
```
Comment (AlexMokin) [inline task/sql-tasks.js:275]: from-clause is optional.

 it's better to write this query like this

```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 13]
```diff
@@ -280,30 +461,49 @@ async function task_1_21(db) {
...
+          ORDER BY OD.UnitPrice DESC, C.CompanyName, P.ProductName;
+  `);
+  return result[0];
 }
 
 module.exports = {
-    task_1_1: task_1_1,
-    task_1_2: task_1_2,
-    task_1_3: task_1_3,
-    task_1_4: task_1_4,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not modify original code

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 13 above)
Comment (Arseni-Lamihov) [reply task/sql-tasks.js]: i don`t understand. i didn`t modify anything

**Comment by Arseni-Lamihov** [comment]
https://travis-ci.com/github/Arseni-Lamihov/db-assignments/builds/182385751

File: task/sql-tasks.js [hunk 14]
```diff
@@ -177,7 +268,13 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            (SELECT COUNT(ProductID)) as 'TotalOfCurrentProducts',
+            (SELECT COUNT(CASE WHEN Discontinued = 1 THEN 1 END )) as 'TotalOfDiscontinuedProducts'
```
Comment (AlexMokin) [inline task/sql-tasks.js]: filter out results using where clause in this subquery

File: task/sql-tasks.js [hunk 15]
```diff
@@ -177,7 +268,13 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+        SELECT
+            (SELECT COUNT(ProductID)) as 'TotalOfCurrentProducts',
+            (SELECT COUNT(CASE WHEN Discontinued = 1 THEN 1 END )) as 'TotalOfDiscontinuedProducts'         
+        FROM Products;
```
Comment (AlexMokin) [inline task/sql-tasks.js]: from-clause is optional

File: task/sql-tasks.js [hunk 16]
```diff
@@ -280,30 +461,49 @@ async function task_1_21(db) {
...
-    task_1_14: task_1_14,
-    task_1_15: task_1_15,
-    task_1_16: task_1_16,
-    task_1_17: task_1_17,
-    task_1_18: task_1_18,
-    task_1_19: task_1_19,
-    task_1_20: task_1_20,
-    task_1_21: task_1_21,
-    task_1_22: task_1_22
+  task_1_1: task_1_1,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: github shows that this lines are changed. Do not modify original code

image

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by Arseni-Lamihov** [comment]
https://travis-ci.com/github/Arseni-Lamihov/db-assignments/builds/182385751

File: task/sql-tasks.js [hunk 17]
```diff
@@ -177,7 +268,14 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+  let result = await db.query(`
+      SELECT
+        (SELECT COUNT(ProductID) FROM Products) as 'TotalOfCurrentProducts',
+        (SELECT COUNT(*) FROM Products WHERE Discontinued = 1) as 'TotalOfDiscontinuedProducts'
+    FROM Products
```
Comment (AlexMokin) [inline task/sql-tasks.js:275]: from-clause is optional

**Review by AlexMokin — COMMENTED** [review]
Accepted, But seriously from clause is optional.

**Pull request closed without merging**